omnigibson/action_primitives/starter_semantic_action_primitives.py

```diff
@@ -1940,32 +1898,62 @@ def _sample_pose_with_object_and_predicate(
+            candidate_poses (list of arrays): Candidate 2d poses (x, y, yaw)
+            pose_on_obj (Iterable of arrays or list of Iterables): Pose(s) on the object(s) in the world frame.
+                Can be a single pose or list of poses.
+            arm (str): Name of the arm to check reachability for
+            skip_obstacle_update (bool): Whether to skip updating the obstacles in the motion generator
 
         Returns:
-            bool: True if the robot is in a valid pose, False otherwise
+            list of bool: Whether any robot arm can reach all poses on the objects and is not in collision
+                at the specified 2d poses
         """
-        pose = self._get_robot_pose_from_2d_pose(pose_2d)
-        if pose_on_obj is not None:
-            relative_pose = T.relative_pose_transform(*pose_on_obj, *pose)
-            if not self._target_in_reach_of_robot_relative(relative_pose):
-                return False
+        if arm is None:
+            arm = self.arm
 
-        if set_base_and_detect_collision(context, pose):
-            indented_print("Candidate position failed collision test.")
-            return False
-        return True
+        # First check collisions for all candidate poses
+        candidate_joint_positions = []
+        current_joint_pos = self.robot.get_joint_positions()
+        for pose in candidate_poses:
+            joint_pos = current_joint_pos.clone()
+            joint_pos[self.robot.base_idx[:2]] = pose[:2]
+            joint_pos[self.robot.base_idx[3:]] = th.tensor([0.0, 0.0, pose[2]])
+            candidate_joint_positions.append(joint_pos)
 
-    @staticmethod
-    def _get_robot_pose_from_2d_pose(pose_2d):
+        candidate_joint_positions = th.stack(candidate_joint_positions)
+
+        # If an object is grasped, we need to pass it to the collision checker
+        obj_in_hand = self._get_obj_in_hand()
+        attached_obj = {self.robot.eef_link_names[self.arm]: obj_in_hand.root_link} if obj_in_hand is not None else None
+
+        invalid_results = self._motion_generator.check_collisions(
+            candidate_joint_positions,
+            self_collision_check=False,
+            skip_obstacle_update=skip_obstacle_update,
+            attached_obj=attached_obj,
+        ).cpu()
+
+        # For each candidate that passed collision check, verify reachability
+        for i in range(len(candidate_poses)):
+            if invalid_results[i].item():
+                continue
+
+            if pose_on_obj is not None:
+                pose = self._get_robot_pose_from_2d_pose(candidate_poses[i])
+                relative_pose = T.relative_pose_transform(*pose_on_obj, *pose)
+                if not self._target_in_reach_of_robot_relative(
+                    relative_pose, skip_obstacle_update=skip_obstacle_update
+                )[arm]:
+                    invalid_results[i] = True
+
+        return ~invalid_results
+
+    def _get_robot_pose_from_2d_pose(self, pose_2d):
```

Comment:
Figure out exactly what this function should return?

Reply:
quat = R.from_euler("XYZ", [cur_joint_rx, cur_joint_ry, cur_joint_rz]).as_quat()